File: LRU_cache.cpp
class LRUCache {
    unordered_map<int, list<pair<int, int>>::iterator> mp;
    list<pair<int, int>> li;
    const int capacity;

public:
    LRUCache(int capacity) : capacity(capacity) {}

    int get(int key) {
        if (!mp.count(key))
            return -1;
        li.push_front(*mp[key]);
        li.erase(mp[key]);
        mp[key] = li.begin();
        return li.begin()->second;
    }

    void put(int key, int value) {
        if (mp.count(key)) {
            li.erase(mp[key]);
        } else if (mp.size() >= capacity) {
            mp.erase(li.back().first);
            li.pop_back();
        }

        li.push_front({key, value});
        mp[key] = li.begin();
    }
};

File: LRU_cache.go
type LRUCache struct {
    m map[int]*list.Element
    list *list.List
    capacity int 
}

type KeyValue struct {
    key int
    value int
}

func NewLRUCache (capacity int) LRUCache {
    var cache LRUCache
    cache.capacity = capacity
    cache.m = make(map[int]*list.Element)
    cache.list = list.New()
    return cache
}

func (lru *LRUCache) Get(key int) int {
    if e, ok := lru.m[key]; ok {
        lru.list.MoveToFront(e)
        return e.Value.(*KeyValue).value
    } 
    
    return -1
}

func (lru *LRUCache) Put(key, value int) {
    if e, ok := lru.m[key]; ok {
        lru.list.MoveToFront(e)
        e.Value = &KeyValue{key, value}
    } else {
        if lru.list.Len() == lru.capacity {
            e := lru.list.Back()
            delete(lru.m, e.Value.(*KeyValue).key)
            lru.list.Remove(e)
        }

        lru.list.PushFront(&KeyValue{key, value})
        lru.m[key] = lru.list.Front()
    }
}
